# Coding and Programming

## Capítulo 6

### Resolución de problemas con algoritmos

- **Unidad 30.** Enfoque Greedy algoritmos voraces o ávidos.
- **Unidad 31.** Divide y vencerás
- **Unidad 32.** Programación dinámica
- **Unidad 33.** Backtracking


___

## Q 30-01

Este es el problema del intercambio de monedas.

Tienes una moneda de 400 wones.

Escriba el resultado de cómo el algoritmo `coin_change()` determina el cambio de la moneda de 800 wones.
```
coins  = [500, 400, 100, 50, 10]
amount = int(input("Ingrese la cantidad: "))
changes= coin_change(coins, amount)
print(changes, len(changes))

Ingrese la cantidad: 800
```

[500, 100, 100, 100] 4

___

## Q 31-01

Tienes 8 monedas idénticas numeradas del 1 al 8. De estas, solo una moneda es más pesada que la otra.

Suponiendo que puede pesar monedas con una balanza equilibrada, diseñe un algoritmo que encuentre la moneda más pesada.

En este momento, ¿cuántas veces se debe usar al menos la balanza?

In [5]:
def mayor_peso(monedero):
    if len(monedero) == 1:
        return monedero[0]
    elif len(monedero) == 2:
        return max(monedero)
    
    division_grupos = len (monedero) // 3
    primer_grupo = monedero[:division_grupos]
    segundo_grupo = monedero[division_grupos:2*division_grupos]
    tercer_grupo = monedero[2*division_grupos:]

    mas_pesada = max(primer_grupo, segundo_grupo, tercer_grupo, key=sum)

    return mayor_peso(mas_pesada)

monedas = [1,2,3,4,5,6,7,8]

pesa_mas = int(mayor_peso(monedas))
print("La más pesada es: ", pesa_mas)

La más pesada es:  8


Al dividir cada grupo en 3 monedas, se pesan dos grupos de 3 monedas en cada balanza, si queda equilibrada, quiere decir que todas pesan lo mismo, por ende, la moneda más pesada está en una de las dos sobrantes. En este caso como maximo se utilizaría dos veces la balanza. Y si la balanza se inclina para un lado, se hace la medicion entre dos de las monedas de ese lado que se inclinó y lo mismo, si se mantiene equilibrada es porque la moneda más pesada fue la que no se colocó a medir.

___

## Q 31-02

Hay nueve monedas idénticas numeradas del 1 al 9 en el problema anterior.

Diseñe un algoritmo para encontrar una moneda pesada en este caso.

En este momento, ¿cuántas veces se debe usar al menos la balanza?

In [ ]:
def mayor_peso(monedero):
    if len(monedero) == 1:
        return monedero[0]
    elif len(monedero) == 2:
        return max(monedero)
    
    division_grupos = len (monedero) // 3
    primer_grupo = monedero[:division_grupos]
    segundo_grupo = monedero[division_grupos:2*division_grupos]
    tercer_grupo = monedero[2*division_grupos:]

    mas_pesada = max(primer_grupo, segundo_grupo, tercer_grupo, key=sum)

    return mayor_peso(mas_pesada)

monedas = [1,2,3,4,5,6,7,8, 9]

pesa_mas = int(mayor_peso(monedas))
print("La más pesada es: ", pesa_mas)

Sigue la misma logica que el anterior, solo que en esta en el peor de los casos, se tendría que utilizar 3 veces dicha balanza.


___

## Q 32-01

Analice el resultado de la ejecución del siguiente código y compárelo con el rendimiento de las funciones `fib1()`, `fib2()` y `fib3()`.
```
def fib4(n):
    if n <= 1:
        return n
    else:
        a,b = 0,1
        for i in range(2, n + 1):
            a,b = b,a+b
        return b

N = int(input("Ingrese un numero: "))
print(fib4(N))


Ingrese un numero:
```


In [8]:
def fib4(n):
    if n <= 1:
        return n
    else:
        a,b = 0,1
        for i in range(2, n + 1):
            a,b = b,a+b
        return b

N = int(input("Ingrese un numero: "))
print(fib4(N))

55


En caso de n ser igual o menor a 1, retorna el valor de N, sino, iguala una variable a igual a 0 y una b a 1, entonces crea un bucle para un iterador en el rango de 2 hasta n + 1, mientras esto se cumpla, a va a tener el valor de b y a b se le va a sumar el valor de a+b, retornando al final, b.

En comparacion de los demás, este brinda mayor eficiencia y simplicidad.

___

## Q 32-02

Analice el resultado de la ejecución del siguiente código y compárelo con el rendimiento de las funciones `bin1()` y `bin2()`.
```
def bin3(n,k):
    B = [0] * (n + 1)
    for i in range(n + 1):
        for j in range(n, -1, -1):
            if j==0 or j==i:
                B[j] = 1
            else:
                B[j] = B[j] B[j-1]
    return B[k]


for i in range(10):
    for j in range(i+1):
        print(bin3(i,j),end=" ")
    print()
```

In [2]:
def bin3(n,k):
    B = [0] * (n + 1)
    for i in range(n + 1):
        for j in range(n, -1, -1):
            if j==0 or j==i:
                B[j] = 1
            else:
                B[j] = B[j] + B[j-1]
    return B[k]


for i in range(10):
    for j in range(i+1):
        print(bin3(i,j),end=" ")
    print()

1 
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 


A diferencia de los anteriores, este toma los valores de i y j en un rango de 10 - 1 en  i y de 10 en J.

Este funciona de manera más efectiva debido a que utiliza for's anidados los cuales al no tener que guardar valores de variables, utilizan menos espacios de memoria.

___

## Q 33-01

Si reemplaza cada letra con un número en la siguiente letra, el número correspondiente a las cinco palabras son todos cuadrados.

Para cada palabra, la suma de cada dígito también es un número cuadrado.

Encuentra el número representado por cada letra.

**Código de referencia:**

In [2]:
from math import sqrt, floor
import time


def is_square_digitsum(n):
    S = 0  # Se inicializa la variable S
    while n > 0:
        S += n % 10  # Se corrige el error, se suma el dígito a la variable S
        n //= 10  # Se corrige el error, se utiliza operador de división entera
    if sqrt(S) == int(sqrt(S)):
        return True
    return False


def find_all_squares():
    sqrs = [[] for _ in range(5)]
    for i in range(1, floor(sqrt(10 ** 5)) + 1):
        n = i * i
        if not is_square_digitsum(n):
            continue
        s = str(n)
        if len(s) == 3 and s[1] != s[2]:
            continue
        if len(s) == 5 and s[2] != s[3]:
            continue
        if len(s) in [4, 5] and len(set(s)) != 4:
            continue
        sqrs[len(s) - 1].append(n)
    return sqrs


def promising(s, n, dic):
    for i in range(len(s)):
        digit = int(str(n)[i])
        for key, value in dic.items():
            if key == s[i] and value != digit:
                return False
            if value == digit and key != s[i]:
                return False
    return True


def solve(words, dic, squares):
    global solved
    if len(words) == 0:
        solved = dic
    else:
        s = words[0]
        candidates = squares[len(s) - 1]
        for n in candidates:
            if promising(s, n, dic):
                newdic = dic.copy()
                for i in range(len(s)):
                    newdic[s[i]] = int(str(n)[i])
                solve(words[1:], newdic, squares)


def main():
    squares = find_all_squares()
    words = ['A', 'TO', 'ALL', 'XMAS', 'MERRY']
    dic = {}
    solve(words, dic, squares)
    for word in words:
        print(word, end=": ")
        for c in word:
            print(solved[c], end="")
        print()


start = time.time()
solved = {}
main()
end = time.time()
print("Elapsed Time:", end - start, "seconds")


A: 9
TO: 81
ALL: 900
XMAS: 7396
MERRY: 34225
Elapsed Time: 0.0 seconds


A: 169
TO: 144
ALL: 961
XMAS: 7225
MERRY: 7396

A: 169 (13^2), donde 1+6+9=16 (4^2)

TO: 144 (12^2), donde 1+4+4=9 (3^2)

ALL: 961 (31^2), donde 9+6+1=16 (4^2)

XMAS: 7225 (85^2), donde 7+2+2+5=16 (4^2)

MERRY: 7396 (86^2), donde 7+3+9+6=25 (5^2)

# Fin